### PR TITLE
Mapping Restructure

### PR DESCRIPTION
**Pull request by timfrazee**
Makes some changes to the structure of the `metadata_mapper` directory.

This is a small refactor upon which future mapper PRs will be based.

* Refines and renames base mapper and vernacular classes
* Adds OAI subdirectory for `oai_mapper` and its descendants
* Sets up mapping class hierarchy to still call `to_UCLDC`, but delegate the act of building the maps to a new `UCLDC_map` method, which goes up the ancestor chain to build the mapped metadata.

Also includes a few minor dev env updates:

* Adds `python-dotenv` for easier environment variable control in dev
* Sets python version for pyenv
* A few changes to more easily manually run mapping from the CLI in dev

### CODE REVIEW

File: metadata_mapper/lambda_function.py [hunk 1]
```diff
@@ -49,34 +53,30 @@ def parse_enrichment_url(enrichment_url):
...
 def map_page(payload, context):
-    if settings.LOCAL_RUN:
+    if settings.LOCAL_RUN and not isinstance(payload, dict):
         payload = json.loads(payload)
 
     vernacular_reader = import_vernacular_reader(payload.get('mapper_type'))
     source_vernacular = vernacular_reader(payload)
     api_resp = source_vernacular.get_api_response()
-    source_metadata_records = source_vernacular.parse(api_resp)
+    mapped_records = source_vernacular.parse(api_resp)
```
Comment (amywieliczka) [inline metadata_mapper/lambda_function.py:62]: These records shouldn't be mapped yet at this point. They should be parsed from the stashed API response, but not yet mapped. There are a few enrichment processes that run prior to mapping.

File: metadata_mapper/lambda_function.py [hunk 2]
```diff
@@ -49,34 +53,30 @@ def parse_enrichment_url(enrichment_url):
...
-    for enrichment_url in collection.get('rikolti__pre_mapping'):
+    for enrichment_url in collection.get('rikolti__pre_mapping', []):
         enrichment_func, kwargs = parse_enrichment_url(enrichment_url)
         if not enrichment_func and settings.SKIP_UNDEFINED_ENRICHMENTS:
             continue
         logging.debug(
             f"[{collection['id']}]: running enrichment: {enrichment_func} "
             f"for page {payload['page_filename']} with kwargs: {kwargs}")
-        source_metadata_records = [
+        mapped_records = [
```
Comment (amywieliczka) [inline metadata_mapper/lambda_function.py:72]: these are the enrichments that get run prior to mapping the record.

File: metadata_mapper/lambda_function.py [hunk 3]
```diff
@@ -49,34 +53,30 @@ def parse_enrichment_url(enrichment_url):
...
             record.enrich(enrichment_func, **kwargs)
-            for record in source_metadata_records
+            for record in mapped_records
         ]
 
-    mapped_records = [record.to_UCLDC() for record in source_metadata_records]
     writer = UCLDCWriter(payload)
-    if settings.DATA_DEST == 'local':
```
Comment (amywieliczka) [inline metadata_mapper/lambda_function.py:75]: We are interested in adding the capacity to see records directly after the mapping step, but before the bulk of the enrichments (excepting those "pre enrichments") have been run. I think this is largely for debug purposes and probably deserves some sort of flag.

File: metadata_mapper/lambda_function.py (hunk 2 above)
Comment (amywieliczka) [reply metadata_mapper/lambda_function.py:72]: perhaps it is worth doing some analysis of these enrichments to see if we can just roll them in with either the parsing logic (in the case of character strippers) or the mapping logic (in the case of id assignments).

File: metadata_mapper/mappers/mapper.py [hunk 4]
```diff
@@ -37,38 +68,49 @@ def get_s3_api_response(self):
...
-        s3_client.put_object(
-            ACL='bucket-owner-full-control',
-            Bucket=bucket,
-            Key=key,
-            Body=json.dumps(mapped_metadata))
+    def __init__(self, collection_id: int, record: dict[str, Any]):
+        self.collection_id: int = collection_id
+        self.source_metadata: dict = record
+        if self.auto_map:
+            self.to_UCLDC()  # By default, generate mapped metadata
```
Comment (amywieliczka) [inline metadata_mapper/mappers/mapper.py:79]: See aforementioned comments regarding pre-mapping enrichments.

File: metadata_mapper/lambda_function.py (hunk 1 above)
Comment (timfrazee) [reply metadata_mapper/lambda_function.py]: Reverted

File: metadata_mapper/lambda_function.py (hunk 2 above)
Comment (timfrazee) [reply metadata_mapper/lambda_function.py]: Reverted

File: metadata_mapper/lambda_function.py (hunk 3 above)
Comment (timfrazee) [reply metadata_mapper/lambda_function.py]: Reverted

File: metadata_mapper/mappers/mapper.py (hunk 4 above)
Comment (timfrazee) [reply metadata_mapper/mappers/mapper.py]: Reverted

File: metadata_mapper/mappers/oai/islandora_mapper.py [hunk 5]
```diff
@@ -1,62 +1,12 @@
...
-import json
-from .mapper import VernacularReader, Record
-from lxml import etree
-from sickle import models
 import requests
-import settings
 
-
-# https://github.com/calisphere-legacy-harvester/dpla-ingestion/blob/ucldc/lib/mappers/islandora_oai_dc_mapper.py
-# https://github.com/ucldc/harvester/blob/master/harvester/fetcher/oai_fetcher.py
```
[the remark this answers is not in the document]
Comment (timfrazee) [reply metadata_mapper/mappers/oai/islandora_mapper.py:11]: Restored

**Comment by timfrazee** [comment]
@amywieliczka @barbarahui The changes and reversions we discussed have been implemented.

This PR does not include any of the upstream changes to OaiMapper or CcaVaultMapper - those will be rebased in their own branches/PRs.

**Review by amywieliczka — APPROVED** [review]

**Comment by lthurston** [comment]
Despite being approved, I'm going to close this and start with a new PR based on a branch that implements some of the mapper changes approved this week. That PR is here: https://github.com/ucldc/rikolti/pull/287

**Pull request closed without merging**